### PR TITLE
 Add support for UTF8 string literals

### LINKED ISSUE #2923: UTF-8 string literals are not mutated
**Describe the bug**
I've noticed that UTF-8 string literals are not mutated.

**Expected behavior**
I'd expect the following line to be mutated by the string literal mutator:

``` cs
var stringLiteral = "mystring"u8;
```

### PR DESCRIPTION
**Pull request by iamdmitrij**
Fixes #2923.

Implemented the same logic as in plain strings:
  - `""u8` -> `"Stryker was here!"u8`
  - `""something8` -> `""u8`

`SyntaxFactory.Literal(string)` method used for string mutations doesn't work with UTF8 string literal type (`ReadOnlySpan<byte>`), so I've taken inspiration on how to make it work from dotnet/roslyn analyzers' [code](https://github.com/dotnet/roslyn/blob/ac301e22853c9083a7aa8f22d49e389056da89de/src/Analyzers/CSharp/CodeFixes/UseUtf8StringLiteral/UseUtf8StringLiteralCodeFixProvider.cs#L193).

Tasks:
- [x] Modify `StringMutator` to support UTF8 string literals
- [x] Unit tests
- [x] Integration tests
- [x] Edit docs

### CODE REVIEW

File: src/Stryker.Core/Stryker.Core/Mutators/StringMutator.cs [hunk 1]
```diff
@@ -17,24 +17,53 @@ public override IEnumerable<Mutation> ApplyMutations(LiteralExpressionSyntax nod
...
+        string replacementValue;
+
+        if (IsStringLiteral(node))
+        {
+            currentValue = (string)node.Token.Value;
+            replacementValue = currentValue == "" ? "Stryker was here!" : "";
+            syntaxNode = SyntaxFactory.LiteralExpression(SyntaxKind.StringLiteralExpression,
+                SyntaxFactory.Literal(replacementValue));
+        }
+        else if(IsUtf8StringLiteral(node))
```
Comment (iamdmitrij) [inline src/Stryker.Core/Stryker.Core/Mutators/StringMutator.cs:36]:
```suggestion
        else if (IsUtf8StringLiteral(node))
```

**Review by richardwerkman — CHANGES_REQUESTED** [review]
Thanks for your first contribution! I've got some suggestions

**Comment by richardwerkman** [comment]
I tested your PR on the following code and it broke:

``` cs
        public string Test()
        {
            return "Hello " + " " + "World";
        }
```

So there is some issue with the mutator logic

File: integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Linq;
+
+namespace ExampleProject.String
+{
+    public class Utf8StringMagic
+    {
+        public ReadOnlySpan<byte> HelloWorld()
+        {
+            return "Hello"u8 + " "u8 + "World!"u8;
```
[the remark this answers is not in the document]
Comment (rouke-broersma) [reply integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs:10]: This should be fixed in this PR, not as a separate defect.

**Comment by iamdmitrij** [comment]
> I tested your PR on the following code and it broke:
> 
> ```cs
>         public string Test()
>         {
>             return "Hello " + " " + "World";
>         }
> ```
> 
> So there is some issue with the mutator logic

Can you specify what exactly fails here?

I have noticed it doesn't compile when concatenation operator `+` is mutated to `-`.

```
var a1 = "Hello"u8 + " "u8 + "World"u8; // OK
var a2 = "Hello"u8 - " "u8 - "World"u8; // Doesn't compile

var a3 = "Hello" + " " + "World"; // OK
var a4 = "Hello" - " " - "World"; // Doesn't compile
```

If that's the case, should be be fixed somehow? Because I don't see how current or previous string mutator code has solution for that. My best guess would be to fix [BinaryExpressionMutator](https://github.com/stryker-mutator/stryker-net/blob/a42634c9a67a565a10aa5c1505f1569cb75ee1c3/src/Stryker.Core/Stryker.Core/Mutators/BinaryExpressionMutator.cs#L51C1-L54C14) implementation to avoid this.

**Comment by richardelekta** [comment]
> I have noticed it doesn't compile when concatenation operator `+` is mutated to `-`.

This isn't the problem, we prevent this mutation from being placed.

```
var a1 = "Hello"u8 + " "u8 + "World"u8; // OK
var a2 = ""u8 + " "u8 + "World"u8; // Doesn't compile

var a3 = "Hello" + " " + "World"; // OK
var a4 = "" + " " + "World"; // OK
```

The above displays the issue. The interesting part is that the code should compile, but it doesn't because of how stryker places the mutation. I haven't investigate yet what exactly goes wrong, but my guess is that there is a flaw in our mutation placing logic

**Comment by iamdmitrij** [comment]
> > I have noticed it doesn't compile when concatenation operator `+` is mutated to `-`.
> 
> This isn't the problem, we prevent this mutation from being placed.
> 
> ```
> var a1 = "Hello"u8 + " "u8 + "World"u8; // OK
> var a2 = ""u8 + " "u8 + "World"u8; // Doesn't compile
> 
> var a3 = "Hello" + " " + "World"; // OK
> var a4 = "" + " " + "World"; // OK
> ```
> 
> The above displays the issue. The interesting part is that the code should compile, but it doesn't because of how stryker places the mutation. I haven't investigate yet what exactly goes wrong, but my guess is that there is a flaw in our mutation placing logic

Thank you. Now it's clear, I will look into this use-case.

**Comment by richardwerkman** [comment]
@iamdmitrij The integration tests have been updated and now also include my previous example.

File: integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamdmitrij) [reply integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs:10]: I've investigated why this test is failing. It fails during mutated code compilation. Stryker by default add ternary expression to each operand:

So this statement
```
return "Hello"u8 + " "u8 + "World!"u8;
```

is transformed into this:
```
return (StrykerqBEu3bP1rxHLxTW.MutantControl.IsActive(99)?""u8 :"Hello"u8 )+ (StrykerqBEu3bP1rxHLxTW.MutantControl.IsActive(100)?""u8 :" "u8 )+ (StrykerqBEu3bP1rxHLxTW.MutantControl.IsActive(101)?""u8:"World!"u8);
```

It boxes each UTF-8 value to `ReadOnlySpan<byte>()` type. The problem lies with C# and how it allows to concatenate UTF-8 string literals. It allows to concatenate UTF-8 string directly: `"Hello"u8 + " "u8 + "World!"u8`, but not when they are represented as `ReadOnlySpan<byte>`: ~~`new ReadOnlySpan<byte>() + new ReadOnlySpan<byte>()`~~.

Hence, mutated code fails to compile with:
```
CS9047	Operator '+' cannot be applied to operands of type 'ReadOnlySpan<byte>' and 'ReadOnlySpan<byte>' that are not UTF-8 byte representations
```

Do you think it's worth fixing this in this PR as it touches core `MutantControl` injection logic? If so, any new ideas are welcome here. So far, I've only thought of converting UTF-8 strings into arrays and concatenating them manually using LINQ when ternary conditions are used:

```
public ReadOnlySpan<byte> HelloWorld()
{
    return (true ? ""u8 : "Hello"u8).ToArray()
        .Concat((true ? ""u8 : " "u8).ToArray()
            .Concat((true ? ""u8 : "World!"u8).ToArray()))
        .ToArray();
}
```

[the remark this answers is not in the document]
Comment (richardwerkman) [reply integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs:10]: Interesting. In that case, we could maybe include some extra code in the compilation that contains an operator overload for `+` on `ReadOnlySpan<byte>`. During the compilation we already pass some custom code to be compiled together with the source project, so that should be fairly easy to do.

> Do you think it's worth fixing this bug in current PR?

Yes, as this seems like an easy fix we should add to this PR. Otherwise, we risk that this will never be fixed once this PR has been merged.

[the remark this answers is not in the document]
Comment (dupdob) [reply integrationtest/TargetProjects/TargetProject/String/Utf8StringMagic.cs:10]: I am not sure adding this operator is needed. 
1. `"Hello"u8 + " "u8 + "World!"u8;` is actually a compile time constant so the compiler concatenate the strings and does not compile any expression. See the actual result:
`    internal static readonly __StaticArrayInitTypeSize=13 5A09E8FA9C77807B24E99C9CF999DEBFAD8441E269EB960E201F61FC3DE20D5A/* Not supported: data(48 65 6C 6C 6F 20 57 6F 72 6C 64 21 00) */;
`
An array of 13 bytes. This is seen as a single string here. As such, it does not really make sense to mutate sub part.
2. In any case, there is no interest in mutating this expression more than once. It is almost certain that if removing `Hello` is killed by a failed test, that very same test will kill removing the space or `world!`. So there is no compelling reason to keep each sub mutation. Note that this remarks is also valid for string concatenation
3. As such, it would make more sense to mutate the whole expression once, which would remove any problems with the missing operator.
4. Adding this operator requires injecting a dedicated `using` statement at the start of any files needing it; meaning it would have to be added everywhere. Furthermore, it could result in compilation errors `ambiguous call` that will require a new roll back logic (to detect it and remove the unnecessary `using` statement.
5. The only consequence of not adding this operand is simply that concatenated u8 strings will not be mutated; which looks alright to me.

TLDR;
I recommend: doing nothing for now and contemplate detecting concatenated constant strings (u8 or otherwise) to be mutated at once. But it would require specific logic within the `ExpressionOrchestrator`
